### PR TITLE
feat!: new authorization flow

### PR DESCRIPTION
**Pull request by Gozala**
Implements https://github.com/web3-storage/w3up/issues/945

- I had to drop the execa and implement my own process spawning observing module because execa was not reliable with interactive test cases
- I also end up dropping ava as it was very difficult to use debugger with it and it was very slow

⚠️ I discovered that agent store is prone to race conditions if you `getClient` more then once, we probably need to fix that but as in a followup

### CODE REVIEW

File: space.js [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+        console.error(result.error)
+      }
+    } else {
+      console.log(
+        `⏭️ Skip account authorization. You can still can manage space using "paper key"`
+      )
+    }
+  }
+
+  console.log(`⁂ ${space.did()}`)
```
Comment (alanshaw) [inline space.js:107]:
```suggestion
  console.log(`⁂ Space created: ${space.did()}`)
```

File: space.js [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+  const recoveryKey = W3Space.toMnemonic(space)
+
+  if (options.caution === false) {
+    console.log(formatRecoveryInstruction(recoveryKey))
+    return space
+  } else {
+    const verified = await mnemonic({
+      secret: recoveryKey.split(/\s+/g),
+      message:
+        'You need to save following "secret recovery key" somewhere safe. For example write it down on piece of paper and put it inside your favorite book.',
```
Comment (alanshaw) [inline space.js:234]:
```suggestion
        'You need to save the following secret recovery key somewhere safe! For example write it down on a piece of paper and put it inside your favorite book.',
```

File: space.js [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+      revealMessage:
+        '🤫 Make sure no one is eavesdropping and hit enter to reveal the key',
+      submitMessage: '📝 Once you have saved the key hit enter to continue',
+      validateMessage:
+        '🔒 Please type or paste recovery key to make sure it is correct',
```
Comment (alanshaw) [inline space.js:239]:
```suggestion
        '🔒 Please type or paste your recovery key to make sure it is correct',
```

File: space.js [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+
+    return verified ? space : null
+  }
+}
+
+/**
+ * @param {string} key
+ */
+const formatRecoveryInstruction = (key) =>
+  `🔑 You need to save following "secret recovery key" somewhere safe. For example write it down on piece of paper and put it inside your favorite book.
```
Comment (alanshaw) [inline space.js:251]:
```suggestion
  `🔑 You need to save following secret recovery key somewhere safe! For example write it down on a piece of paper and put it inside your favorite book.
```

File: space.js [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+/**
+ * @param {string} name
+ * @param {{name:string}[]} spaces
+ * @returns {Promise<string>}
+ */
+const chooseName = async (name, spaces) => {
+  const space = spaces.find((space) => String(space.name) === name)
+  const message =
+    name === ''
+      ? 'What would you like to call this space ?'
```
Comment (alanshaw) [inline space.js:266]:
```suggestion
      ? 'What would you like to call this space?'
```

File: space.js [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+      : space
+      ? `Name "${space.name}" is already taken, choose another`
```
Comment (alanshaw) [inline space.js:268]:
```suggestion
      ? `Name "${space.name}" is already taken, please choose a different one`
```

File: space.js [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+    return chooseAccount(accounts)
+  }
+}
+
+/**
+ * @param {import('@web3-storage/w3up-client').Client} client
+ */
+export const setupAccount = async (client) => {
+  const email = await input({
+    message: `📧 Please enter email address to setup an account`,
```
Comment (alanshaw) [inline space.js:330]:
```suggestion
    message: `📧 Please enter an email address to setup an account`,
```

File: test/helpers/process.js [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+//     for await (const chunk of source) {
+//       writer.write(chunk)
+//     }
+//     await writer.close()
+//   } catch (error) {
+//     await writer.abort(error)
+//   } finally {
+//     writer.releaseLock()
+//   }
+// }
```
Comment (alanshaw) [inline test/helpers/process.js:389]: Remove all the commented?

**Review by alanshaw — APPROVED** [review]
Looks great. Non blocking suggestions left.

File: dialog.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+/**
+ * @param {string} input
+ */
+const parseInput = (input) => input.trim().split(/[\n\s]+/)
+
+/**
+ * @param {string[]} key
+ * @param {string[]} input
```
Comment (Gozala) [inline dialog.js:105]:
```suggestion
 * @param {string[]} input
 * @param {string[]} key
```

File: test/helpers/process.js [hunk 10]
```diff
@@ -0,0 +1,402 @@
+import Process from 'node:child_process'
+import { TextDecoder } from 'node:util'
+import { ByteStream } from './stream.js'
+
+/**
+ * @typedef {object} Command
+ * @property {string} model.program
+ * @property {string[]} model.args
+ * @property {Record<string, string|undefined>} model.env
```
Comment (Gozala) [inline test/helpers/process.js:9]:
```suggestion
 * @property {string} program
 * @property {string[]} args
 * @property {Record<string, string|undefined>} env
```

File: test/helpers/process.js (hunk 8 above)
Comment (Gozala) [inline test/helpers/process.js:389]:
```suggestion
```

File: test/bin.spec.js [hunk 11]
```diff
@@ -1,862 +1,1125 @@
...
+  }),
+}
+
+export const testSpace = {
+  'w3 space create': test(async (assert, context) => {
+    const command = w3.args(['space', 'create']).env(context.env.alice).fork()
+
+    const line = await command.output.take(1).text()
+
+    assert.match(line, /What would you like to call this space \?/)
```
Comment (Gozala) [inline test/bin.spec.js:99]:
```suggestion
    assert.match(line, /What would you like to call this space/)
```

File: test/bin.spec.js [hunk 12]
```diff
@@ -1,862 +1,1125 @@
...
+          '--no-recovery',
+          '--customer',
+          'unknown@web.mail',
+          '--no-account',
+        ])
+        .join()
+        .catch()
+
+      assert.match(output, /billing account/)
+      assert.match(output, /Skip billing setup/)
```
Comment (Gozala) [inline test/bin.spec.js:218]:
```suggestion
      assert.match(output, /Skipped billing setup/)
```

**Pull request merged**